File: src/render/transitions/flash-black-concat.ts
import path from 'node:path';
import { execFileSync } from 'node:child_process';
import { RenderError } from '../../errors.js';
import { updateProgress } from '../task.js';
import type { QualityPreset, ResolutionPreset } from '../../presets.js';
import type { RenderTask } from '../task.js';
import type { VideoClip } from '../types.js';
import {
  collectClips,
  normalizeClips,
  type TransitionHandlerArgs,
  type TransitionHandlerResult,
} from './shared.js';

export function ffmpegFlashBlackConcat(
  ffmpegPath: string,
  clips: VideoClip[],
  outputPath: string,
  qualPreset: QualityPreset,
  task: RenderTask,
  resPreset: ResolutionPreset,
  transitionDuration: number,
): void {
  const n = clips.length;
  const halfD = transitionDuration / 2;
  const fps = resPreset.fps;
  const formatSeconds = (value: number) => value.toFixed(6);

  const inputArgs: string[] = [];
  for (const clip of clips) {
    inputArgs.push('-i', clip.src);
  }

  if (n === 1) {
    const filterComplex = `[0:v]setpts=PTS-STARTPTS[vout]`;
    execFileSync(
      ffmpegPath,
      [
        ...inputArgs,
        '-filter_complex', filterComplex,
        '-map', '[vout]',
        '-c:v', 'libx264',
        '-preset', qualPreset.ffmpegPreset,
        '-crf', String(qualPreset.crf),
        '-pix_fmt', 'yuv420p',
        '-an',
        '-y', outputPath,
      ],
      { timeout: 600_000 },
    );
    updateProgress(task, 1);
    return;
  }

  const filterParts: string[] = [];

  for (let i = 0; i < n; i++) {
    const clip = clips[i];
    const duration = clip.duration;
    const isFirst = i === 0;
    const isLast = i === n - 1;

    const effectiveHalfD = Math.min(halfD, duration / 2);
    if (effectiveHalfD < halfD) {
      console.warn(
        `  ⚠ 片段 ${clip.key} 时长 ${duration.toFixed(2)}s 不足，淡黑时长已自动收缩至 ${(effectiveHalfD * 2).toFixed(2)}s`,
      );
    }

    const fadeOutStart = Math.max(0, duration - effectiveHalfD);
    let chain = `[${i}:v]setpts=PTS-STARTPTS`;

    if (!isFirst) {
      // fade in from black at the start
      chain += `,fade=t=in:st=0:d=${formatSeconds(effectiveHalfD)}`;
    }
    if (!isLast) {
      // fade out to black at the end
      chain += `,fade=t=out:st=${formatSeconds(fadeOutStart)}:d=${formatSeconds(effectiveHalfD)}`;
    }

    chain += `[v${i}]`;
    filterParts.push(chain);
  }

  const concatInputs = Array.from({ length: n }, (_, i) => `[v${i}]`).join('');
  filterParts.push(`${concatInputs}concat=n=${n}:v=1:a=0[vout]`);

  const filterComplex = filterParts.join(';');

  console.log(`\n[2/3] FFmpeg 闪黑拼接 ${n} 个片段（转场 ${transitionDuration}s，每侧 ${halfD}s）...`);
  execFileSync(
    ffmpegPath,
    [
      ...inputArgs,
      '-filter_complex', filterComplex,
      '-map', '[vout]',
      '-c:v', 'libx264',
      '-preset', qualPreset.ffmpegPreset,
      '-crf', String(qualPreset.crf),
      '-pix_fmt', 'yuv420p',
      '-an',
      '-y', outputPath,
    ],
    { timeout: 600_000 },
  );

  updateProgress(task, 1);
}

export function handleFlashBlackConcat(
  args: TransitionHandlerArgs,
): TransitionHandlerResult {
  const { rootDir, ffmpegPath, ffprobePath, request, videoInfo, qualPreset, resPreset, task, outDir, outFile } = args;

  void ffprobePath;

  console.log(`\n[1/3] 收集并校验视频片段...`);

  const transitionDuration = typeof request.variables.transition_duration === 'number'
    ? request.variables.transition_duration
    : 0.5;

  const clips = collectClips(request.variables, request.templateInfo, videoInfo);
  if (clips.length === 0) {
    throw new RenderError('没有可用的视频片段（未提供任何视频）');
  }

  console.log(`  共 ${clips.length} 个片段，闪黑转场时长 ${transitionDuration}s（每侧 ${(transitionDuration / 2).toFixed(2)}s）`);

  const normalized = normalizeClips(
    rootDir,
    ffmpegPath,
    clips,
    qualPreset,
    resPreset,
  );

  const outputPath = path.join(outDir, outFile);
  ffmpegFlashBlackConcat(
    ffmpegPath,
    normalized.clips,
    outputPath,
    qualPreset,
    task,
    resPreset,
    transitionDuration,
  );

  return {
    cleanup: normalized.cleanup,
    outputPath,
  };
}

File: videocut/render/transitions/flash_black_concat.py
from __future__ import annotations

from pathlib import Path

from videocut.errors import RenderError
from videocut.ffmpeg_config import FFmpegVideoSettings
from videocut.log import get_logger
from videocut.presets import QualityPreset, ResolutionPreset
from videocut.render.task import RenderTask, update_progress
from videocut.render.transitions.shared import (
    TransitionHandlerArgs,
    TransitionHandlerResult,
    _run_ffmpeg,
    collect_clips,
    normalize_clips,
)
from videocut.render.types import VideoClip


def ffmpeg_flash_black_concat(
    ffmpeg_path: str,
    clips: list[VideoClip],
    output_path: str,
    qual_preset: QualityPreset,
    video_settings: FFmpegVideoSettings,
    task: RenderTask,
    res_preset: ResolutionPreset,
    transition_duration: float,
) -> None:
    clip_count = len(clips)
    half_duration = transition_duration / 2
    fps = res_preset.fps
    def format_seconds(value: float) -> str:
        return f"{value:.6f}"
    input_args: list[str] = []
    for clip in clips:
        input_args.extend(["-i", clip.src])

    if clip_count == 1:
        _run_ffmpeg(
            [
                ffmpeg_path,
                *video_settings.input_args(),
                *input_args,
                "-filter_complex",
                "[0:v]setpts=PTS-STARTPTS[vout]",
                "-map",
                "[vout]",
                *video_settings.output_args(qual_preset),
                "-an",
                "-y",
                output_path,
            ],
            timeout=600,
        )
        update_progress(task, 1.0)
        return

    filter_parts: list[str] = []
    for index, clip in enumerate(clips):
        duration = clip.duration
        is_first = index == 0
        is_last = index == clip_count - 1
        effective_half = min(half_duration, duration / 2)
        if effective_half < half_duration:
            logger.warning(
                "%s duration %.2fs too short, shrink flash-black to %.2fs",
                clip.key, duration, effective_half * 2,
            )
        fade_out_start = max(0.0, duration - effective_half)
        chain = f"[{index}:v]setpts=PTS-STARTPTS"
        if not is_first:
            chain += f",fade=t=in:st=0:d={format_seconds(effective_half)}"
        if not is_last:
            chain += f",fade=t=out:st={format_seconds(fade_out_start)}:d={format_seconds(effective_half)}"
        chain += f"[v{index}]"
        filter_parts.append(chain)

    concat_inputs = "".join(f"[v{index}]" for index in range(clip_count))
    filter_parts.append(f"{concat_inputs}concat=n={clip_count}:v=1:a=0[vout]")
    logger.info("[2/3] FFmpeg flash-black concat %d clips (transition %.1fs, half %.2fs)...", clip_count, transition_duration, half_duration)
    _run_ffmpeg(
        [
            ffmpeg_path,
            *video_settings.input_args(),
            *input_args,
            "-filter_complex",
            ";".join(filter_parts),
            "-map",
            "[vout]",
            *video_settings.output_args(qual_preset),
            "-an",
            "-y",
            output_path,
        ],
        timeout=600,
    )
    update_progress(task, 1.0)


logger = get_logger(__name__)


def handle_flash_black_concat(args: TransitionHandlerArgs) -> TransitionHandlerResult:
    logger.info("[1/3] Collecting input clips...")
    transition_duration = (
        float(args.request.variables["transition_duration"])
        if isinstance(args.request.variables.get("transition_duration"), (int, float))
        else 0.5
    )
    clips = collect_clips(args.request.variables, args.request.template_info, args.video_info)
    if not clips:
        raise RenderError("No usable video clips were provided.")
    logger.info("  %d clip(s), flash-black transition %.1fs (each half %.2fs)", len(clips), transition_duration, transition_duration / 2)
    normalized_clips, cleanup = normalize_clips(
        args.root_dir,
        args.ffmpeg_path,
        clips,
        args.qual_preset,
        args.res_preset,
        args.video_settings,
    )
    output_path = str(Path(args.out_dir) / args.out_file)
    ffmpeg_flash_black_concat(
        args.ffmpeg_path,
        normalized_clips,
        output_path,
        args.qual_preset,
        args.video_settings,
        args.task,
        args.res_preset,
        transition_duration,
    )
    return TransitionHandlerResult(cleanup=cleanup, output_path=output_path)
